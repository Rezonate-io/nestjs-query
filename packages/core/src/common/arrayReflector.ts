import { Reflector } from './reflector';
import { Class } from './class.type';
import { getClassMetadata, MetaValue } from './reflect.utils';

export class ArrayReflector extends Reflector {
  append<DTO, Data>(DTOClass: Class<DTO>, data: Data): void {
    const metadata = getClassMetadata<DTO, Data[]>(DTOClass, this.metaKey, false) ?? [];
    metadata.push(data);
    this.defineMetadata(metadata, DTOClass);
  }

  get<DTO, Data>(DTOClass: Class<DTO>, includeParents = false): MetaValue<Data[]> {
    return this.getMetadata(DTOClass, includeParents);
  }
}