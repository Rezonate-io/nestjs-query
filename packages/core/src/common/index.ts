export * from './class.type';
export * from './class.utils';
export * from './deep-partial.type';
export * from './misc.utils';
export * from './reflect.utils';
export { Reflector } from './reflector';
export { ValueReflector } from './valueReflector';
export { ArrayReflector } from './arrayReflector';
export { MapReflector } from './mapReflector';
