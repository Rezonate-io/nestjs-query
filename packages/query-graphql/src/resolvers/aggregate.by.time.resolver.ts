// eslint-disable-next-line max-classes-per-file
import { Args, ArgsType, Resolver } from '@nestjs/graphql';
import {
  AggregateByTimeResponse,
  AggregateQuery,
  AggregateResponse,
  Class,
  Filter,
  mergeFilter,
  QueryService,
} from '@rezonate/nestjs-query-core';
import omit from 'lodash.omit';

import { OperationGroup } from '../auth';
import { getDTONames } from '../common';
import {  AuthorizerFilter, ResolverQuery } from '../decorators';
import { AuthorizerInterceptor } from '../interceptors';
import { transformAndValidate } from './helpers';
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface';
import { AggregateByTimeResponseType } from '../types/aggregate/aggregate-by-time-response.type';
import { AggregateByTimeArgsType } from '../types/aggregate/aggregate-by-time-args.type';
import { AggregateResolverOpts } from './aggregate.resolver';
import { AggregateByTimeQueryParam } from '../decorators/aggregate-by-time-query-param.decorator';

export interface AggregateByTimeResolver<DTO, QS extends QueryService<DTO, unknown, unknown>> extends ServiceResolver<DTO, QS> {
  aggregateByTime(
    filter: AggregateByTimeArgsType<DTO>,
    aggregateQuery: AggregateQuery<DTO>,
    authFilter?: Filter<DTO>
  ): Promise<AggregateResponse<DTO>[]>
}

/**
 * @internal
 * Mixin to add `aggregate` graphql endpoints.
 */
export const AggregateableByTime =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts?: AggregateResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<AggregateByTimeResolver<DTO, QS>> & B => {
    if (!opts || !opts.enabled) {
      return BaseClass as never;
    }

    const { baseNameLower } = getDTONames(DTOClass);
    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection');
    const queryName = `${baseNameLower}AggregateByTime`;
    const AR = AggregateByTimeResponseType(DTOClass);

    @ArgsType()
    class AA extends AggregateByTimeArgsType(DTOClass) {}

    @Resolver(() => AR, { isAbstract: true })
    class AggregateByTimeResolverBase extends BaseClass {
      @ResolverQuery(
        () => [AR],
        { name: queryName },
        commonResolverOpts,
        { interceptors: [AuthorizerInterceptor(DTOClass)] },
        opts ?? {},
      )
      async aggregateByTime(
        @Args({ type: () => AA }) args: AA,
        @AggregateByTimeQueryParam() query: AggregateQuery<DTO>,
        @AuthorizerFilter({
          operationGroup: OperationGroup.AGGREGATE,
          many: true,
        })
        authFilter?: Filter<DTO>,
      ): Promise<AggregateByTimeResponse<DTO>> {
        const qa = await transformAndValidate(AA, args);
        const filter = mergeFilter(qa.filter || {}, authFilter ?? {});

        return this.service.aggregateByTime(
          filter,
          query,
          args.field,
          args.from,
          args.to ?? new Date(),
          args.interval.count,
          args.interval.span,
          args.accumulate,
          qa.groupByLimit,
          opts.maxRowsForAggregate,
          opts.maxRowsForAggregateWithIndex,
          opts.limitAggregateByTableSize,
        );
      }
    }

    return AggregateByTimeResolverBase;
  };
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const AggregateByTimeResolver = <
  DTO,
  QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>,
>(
  DTOClass: Class<DTO>,
  opts?: AggregateResolverOpts,
): ResolverClass<DTO, QS, AggregateByTimeResolver<DTO, QS>> => AggregateableByTime(DTOClass, opts)(BaseServiceResolver);
