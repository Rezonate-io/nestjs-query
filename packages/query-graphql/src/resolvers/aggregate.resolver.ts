import { Args, ArgsType, Resolver } from '@nestjs/graphql'
import { AggregateQuery, AggregateResponse, Class, Filter, mergeFilter, QueryService } from '@rezonate/nestjs-query-core'
import omit from 'lodash.omit'

import { OperationGroup } from '../auth'
import { getDTONames } from '../common'
import { AggregateQueryParam, AuthorizerFilter, ResolverMethodOpts, ResolverQuery } from '../decorators'
import { AuthorizerInterceptor } from '../interceptors'
import { AggregateArgsType, AggregateResponseType } from '../types'
import { transformAndValidate } from './helpers'
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface'

export type AggregateResolverOpts = {
  enabled?: boolean
  maxRowsForAggregate?: number
  maxRowsForAggregateWithIndex?: number
} & ResolverMethodOpts

export interface AggregateResolver<DTO, QS extends QueryService<DTO, unknown, unknown>> extends ServiceResolver<DTO, QS> {
  aggregate(
    filter: AggregateArgsType<DTO>,
    aggregateQuery: AggregateQuery<DTO>,
    authFilter?: Filter<DTO>
  ): Promise<AggregateResponse<DTO>[]>
}

/**
 * @internal
 * Mixin to add `aggregate` graphql endpoints.
 */
export const Aggregateable =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts?: AggregateResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<AggregateResolver<DTO, QS>> & B => {
    if (!opts || !opts.enabled) {
      return BaseClass as never
    }

    const { baseNameLower, baseName } = getDTONames(DTOClass)
    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection')
    const queryName = `${baseNameLower}Aggregate`
    const AR = AggregateResponseType(DTOClass)

    @ArgsType()
    class AA extends AggregateArgsType(DTOClass) {}

    @Resolver(() => AR, { isAbstract: true })
    class AggregateResolverBase extends BaseClass {
      @ResolverQuery(
        () => [AR],
        { name: queryName },
        commonResolverOpts,
        { interceptors: [AuthorizerInterceptor(DTOClass)] },
        opts ?? {}
      )
      async aggregate(
        @Args({ type: () => AA }) args: AA,
        @AggregateQueryParam() query: AggregateQuery<DTO>,
        @AuthorizerFilter({
          operationGroup: OperationGroup.AGGREGATE,
          many: true
        })
        authFilter?: Filter<DTO>
      ): Promise<AggregateResponse<DTO>[]> {
        const qa = await transformAndValidate(AA, args)
        const filter = mergeFilter(qa.filter || {}, authFilter ?? {})

        return this.service.aggregate(filter, query, qa.groupByLimit, opts.maxRowsForAggregate, opts.maxRowsForAggregateWithIndex)
      }
    }

    return AggregateResolverBase
  }
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const AggregateResolver = <DTO, QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>>(
  DTOClass: Class<DTO>,
  opts?: AggregateResolverOpts
): ResolverClass<DTO, QS, AggregateResolver<DTO, QS>> => Aggregateable(DTOClass, opts)(BaseServiceResolver)
