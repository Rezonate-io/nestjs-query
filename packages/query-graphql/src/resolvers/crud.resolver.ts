import { Resolver } from '@nestjs/graphql';
import { Class, DeepPartial, QueryService } from '@rezonate/nestjs-query-core';

import { mergeBaseResolverOpts } from '../common';
import { BaseResolverOptions } from '../decorators/resolver-method.decorator';
import { ConnectionOptions, PagingStrategies } from '../types';
import { AggregateableByTime } from './aggregate.by.time.resolver';
import { Aggregateable, AggregateResolver, AggregateResolverOpts } from './aggregate.resolver';
import { Creatable, CreateResolver, CreateResolverOpts } from './create.resolver';
import { Deletable, DeleteResolver, DeleteResolverOpts } from './delete.resolver';
import { Readable, ReadResolverFromOpts, ReadResolverOpts } from './read.resolver';
import { Referenceable, ReferenceResolverOpts } from './reference.resolver';
import { Relatable } from './relations';
import { RelatableOpts } from './relations/relations.resolver';
import { BaseServiceResolver, MergePagingStrategyOpts } from './resolver.interface';
import { Updateable, UpdateResolver, UpdateResolverOpts } from './update.resolver';

export interface CRUDResolverOpts<
	DTO,
	R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
	PS extends PagingStrategies = PagingStrategies.CURSOR,
> extends BaseResolverOptions,
	Pick<ConnectionOptions, 'enableTotalCount'> {
	/**
	 * The DTO that should be used as input for create endpoints.
	 */
	CreateDTOClass?: Class<DeepPartial<DTO>>;
	/**
	 * The DTO that should be used as input for update endpoints.
	 */
	UpdateDTOClass?: Class<DeepPartial<DTO>>;
	enableSubscriptions?: boolean;
	pagingStrategy?: PS;
	enableAggregate?: boolean;
	create?: CreateResolverOpts<DTO>;
	read?: R;
	update?: UpdateResolverOpts<DTO>;
	delete?: DeleteResolverOpts<DTO>;
	referenceBy?: ReferenceResolverOpts;
	aggregate?: AggregateResolverOpts;
	limitAggregateByTableSize?: boolean;
}

export interface CRUDResolver<
	DTO,
	R extends ReadResolverOpts<DTO>,
	QS extends QueryService<DTO> = QueryService<DTO>,
> extends CreateResolver<DTO, QS>,
	ReadResolverFromOpts<DTO, R, QS>,
	UpdateResolver<DTO, QS>,
	DeleteResolver<DTO, QS>,
	AggregateResolver<DTO, QS> {
}

function extractRelatableOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): RelatableOpts {
	const { enableTotalCount, enableAggregate, aggregate } = opts;
	return mergeBaseResolverOpts<RelatableOpts>({ enableAggregate, enableTotalCount, ...aggregate }, opts);
}

function extractAggregateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): AggregateResolverOpts {
	const { enableAggregate, aggregate } = opts;
	return mergeBaseResolverOpts<AggregateResolverOpts>({ enabled: enableAggregate, ...aggregate }, opts);
}

function extractCreateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): CreateResolverOpts<DTO> {
	const { CreateDTOClass, enableSubscriptions, create } = opts;
	return mergeBaseResolverOpts<CreateResolverOpts<DTO>>({ CreateDTOClass, enableSubscriptions, ...create }, opts);
}

function extractReadResolverOpts<DTO, R extends ReadResolverOpts<DTO>, PS extends PagingStrategies>(
	opts: CRUDResolverOpts<DTO, R, PagingStrategies>,
): MergePagingStrategyOpts<DTO, R, PS> {
	const { enableTotalCount, pagingStrategy, read } = opts;
	return mergeBaseResolverOpts({
		enableTotalCount,
		pagingStrategy, ...read,
	} as MergePagingStrategyOpts<DTO, R, PS>, opts);
}

function extractUpdateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): UpdateResolverOpts<DTO> {
	const { UpdateDTOClass, enableSubscriptions, update } = opts;
	return mergeBaseResolverOpts<UpdateResolverOpts<DTO>>({ UpdateDTOClass, enableSubscriptions, ...update }, opts);
}

function extractDeleteResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): DeleteResolverOpts<DTO> {
	const { enableSubscriptions, delete: deleteArgs } = opts;
	return mergeBaseResolverOpts<DeleteResolverOpts<DTO>>({ enableSubscriptions, ...deleteArgs }, opts);
}

/**
 * Factory to create a resolver that includes all CRUD methods from [[CreateResolver]], [[ReadResolver]],
 * [[UpdateResolver]], and [[DeleteResolver]].
 *
 * ```ts
 * import { CRUDResolver } from '@rezonate/nestjs-query-graphql';
 * import { Resolver } from '@nestjs/graphql';
 * import { TodoItemDTO } from './dto/todo-item.dto';
 * import { TodoItemService } from './todo-item.service';
 *
 * @Resolver()
 * export class TodoItemResolver extends CRUDResolver(TodoItemDTO) {
 *   constructor(readonly service: TodoItemService) {
 *     super(service);
 *   }
 * }
 * ```
 * @param DTOClass - The DTO Class that the resolver is for. All methods will use types derived from this class.
 * @param opts - Options to customize the resolver.
 */
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const CRUDResolver = <
	DTO,
	R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
	PS extends PagingStrategies = PagingStrategies.CURSOR,
>(
	DTOClass: Class<DTO>,
	opts: CRUDResolverOpts<DTO, R, PS> = {},
):Class<CRUDResolver<DTO, R, any>> => {

	@Resolver(() => DTOClass, { isAbstract: true })
	class BaseResolver extends BaseServiceResolver<DTO, any> {
	}

	const Mixins = [] as ((cls: Class<unknown>) => Class<unknown>)[];

	Mixins.push(Referenceable(DTOClass, opts.referenceBy ?? {}));
	Mixins.push(Relatable(DTOClass, extractRelatableOpts(opts)));
	if (opts.aggregate?.disabled) Mixins.push(Aggregateable(DTOClass, extractAggregateResolverOpts(opts)));
	if (opts.aggregate?.disabled) Mixins.push(AggregateableByTime(DTOClass, extractAggregateResolverOpts(opts)));
	if (opts.create?.disabled) Mixins.push(Creatable(DTOClass, extractCreateResolverOpts(opts)));
	if (opts.read?.disabled) Mixins.push(Readable(DTOClass, extractReadResolverOpts(opts)));
	if (opts.update?.disabled) Mixins.push(Updateable(DTOClass, extractUpdateResolverOpts(opts)));
	if (opts.delete?.disabled) Mixins.push(Deletable(DTOClass, extractDeleteResolverOpts(opts)));

	return Mixins.reduce(
		(CurrResolver, action) => action(CurrResolver),
		BaseResolver,
	) as Class<CRUDResolver<DTO, R, any>>;
};
