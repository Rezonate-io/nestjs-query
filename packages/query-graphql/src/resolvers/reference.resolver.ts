import { BadRequestException } from '@nestjs/common';
import { Resolver, ResolveReference } from '@nestjs/graphql';
import { Class, QueryService } from '@rezonate/nestjs-query-core';

import { getDTONames } from '../common';
import { RepresentationType } from '../federation';
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface';

export interface ReferenceResolverOpts {
  key?: string
}

export interface ReferenceResolverType<DTO, QS extends QueryService<DTO> = QueryService<DTO>> extends ResolverClass<DTO, QS, ServiceResolver<DTO, QS>> {
	resolveReference(representation: RepresentationType):Promise<DTO>;
}

/**
 * @internal
 * Mixin to expose `resolveReference` for a DTO on the resolver.
 */
export const Referenceable =
  <DTO, QS extends QueryService<DTO>>(DTOClass: Class<DTO>, opts: ReferenceResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): B & Class<ReferenceResolverType<DTO, QS>> => {
    if (!('key' in opts) || opts.key === undefined) {
      return BaseClass as B & Class<ReferenceResolverType<DTO, QS>>;
    }
    const { key } = opts;

    @Resolver(() => DTOClass, { isAbstract: true })
    class ResolveReferenceResolverBase extends BaseClass {
      @ResolveReference()
      async resolveReference(representation: RepresentationType): Promise<DTO> {
        const id = representation[key];
        if (id === undefined) {
          throw new BadRequestException(
            `Unable to resolve reference, missing required key ${key} for ${getDTONames(DTOClass).baseName}`,
          );
        }
        return this.service.getById(representation[key] as string | number);
      }
    }

    return ResolveReferenceResolverBase as B & Class<ReferenceResolverType<DTO, QS>>;
  };

export const ReferenceResolver = <DTO, QS extends QueryService<DTO> = QueryService<DTO>>(
  DTOClass: Class<DTO>,
  opts: ReferenceResolverOpts = {},
): Class<ReferenceResolverType<DTO, QS>> => Referenceable<DTO, QS>(DTOClass, opts)(BaseServiceResolver);
