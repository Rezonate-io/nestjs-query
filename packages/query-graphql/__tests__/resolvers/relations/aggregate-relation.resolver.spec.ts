import { Query, Resolver } from '@nestjs/graphql';
import { AggregateQuery, AggregateResponse, Filter } from '@rezonate/nestjs-query-core';
import { deepEqual, objectContaining, when } from 'ts-mockito';

import { AggregateRelationsResolver } from '../../../src/resolvers/relations';
import { AggregateRelationsResolverOpts } from '../../../src/resolvers/relations/aggregate-relations.resolver';
import {
	createResolverFromNest,
	generateSchema,
	TestRelationDTO,
	TestResolverDTO,
	TestService,
} from '../../__fixtures__';
import { Inject } from '@nestjs/common';

describe('AggregateRelationsResolver', () => {
	const expectResolverSDL = async (opts?: AggregateRelationsResolverOpts) => {
		@Resolver(() => TestResolverDTO)
		class TestSDLResolver extends AggregateRelationsResolver(TestResolverDTO, opts ?? {}) {
			@Query(() => TestResolverDTO)
			test(): TestResolverDTO {
				return { id: '1', stringField: 'foo' };
			}
		}

		const schema = await generateSchema([TestSDLResolver]);
		expect(schema).toMatchSnapshot();
	};

	it('should not add read methods if one and many are empty', () => expectResolverSDL());
	describe('aggregate', () => {
		it('should use the object type name', () =>
			expectResolverSDL({ enableAggregate: true, many: { relations: { DTO: TestRelationDTO } } }));

		it('should use the dtoName if provided new test', () =>
			expectResolverSDL({
				enableAggregate: true,
				many: { relations: { DTO: TestRelationDTO, dtoName: 'Test' } },
			}));

		it('should not add read methods if enableAggregate is not true new test', () =>
			expectResolverSDL({ many: { relations: { DTO: TestRelationDTO, disableRead: true } } }));

		describe('aggregate query', () => {
			it('should call the service aggregateRelations with the provided dto', async () => {
				@Resolver(() => TestResolverDTO)
				class TestResolver extends AggregateRelationsResolver(TestResolverDTO, {
					enableAggregate: true,
					one: {
						relation: { DTO: TestRelationDTO },
						custom: { DTO: TestRelationDTO, relationName: 'other' },
					},
					many: {
						relations: { DTO: TestRelationDTO },
						customs: { DTO: TestRelationDTO, relationName: 'others' },
					},
				}) {
					constructor(@Inject() service: TestService) {
						super(service);
					}
				}

				interface FinalizedTestResolver {
					aggregateRelations(dto: TestResolverDTO, query: { filter: Filter<TestRelationDTO> }, aggregate:any, filter:any): any;
				}

				const {
					resolver,
					mockService,
				} = await createResolverFromNest<TestResolver, FinalizedTestResolver>(TestResolver);
				const dto: TestResolverDTO = {
					id: 'id-1',
					stringField: 'foo',
				};
				const filter: Filter<TestRelationDTO> = { id: { eq: 'id-2' } };
				const aggregateQuery: AggregateQuery<TestRelationDTO> = {
					count: ['id'],
					sum: ['testResolverId'],
				};
				const output: AggregateResponse<TestRelationDTO>[] = [
					{
						count: { id: 10 },
						sum: { testResolverId: 100 },
					},
				];
				when(
					mockService.aggregateRelations(
						TestRelationDTO,
						'relations',
						deepEqual([dto]),
						objectContaining(filter),
						objectContaining(aggregateQuery),
						undefined,
						undefined,
						undefined,
						undefined,
					),
				).thenResolve(new Map([[dto, output]]));
				const result = await resolver.aggregateRelations(dto, { filter }, aggregateQuery, {});
				return expect(result).toEqual(output);
			});
		});
	});
});
