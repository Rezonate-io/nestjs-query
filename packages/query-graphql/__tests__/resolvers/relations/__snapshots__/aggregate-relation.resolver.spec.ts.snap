// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateRelationsResolver aggregate should not add read methods if enableAggregate is not true new test 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
  relationsAggregate(
    \\"\\"\\"Filter to find records to aggregate on\\"\\"\\"
    filter: TestRelationDTOFilter

    \\"\\"\\"Limit the number of results group by aggregation can return\\"\\"\\"
    groupByLimit: Float
  ): [TestResolverDTORelationsAggregateResponse!]!
  testsAggregate(
    \\"\\"\\"Filter to find records to aggregate on\\"\\"\\"
    filter: TestRelationDTOFilter

    \\"\\"\\"Limit the number of results group by aggregation can return\\"\\"\\"
    groupByLimit: Float
  ): [TestResolverDTOTestsAggregateResponse!]!
}

input TestRelationDTOFilter {
  freeTextQuery: String
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
  contains: ID
  notContains: ID
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

type TestResolverDTORelationsAggregateGroupBy {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsCountAggregate {
  id: Int
  testResolverId: Int
}

type TestResolverDTORelationsMinMaxAggregate {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsAggregateResponse {
  groupBy: TestResolverDTORelationsAggregateGroupBy
  count: TestResolverDTORelationsCountAggregate
  distinctCount: TestResolverDTORelationsCountAggregate
  min: TestResolverDTORelationsMinMaxAggregate
  max: TestResolverDTORelationsMinMaxAggregate
}

type TestResolverDTOTestsAggregateGroupBy {
  id: ID
  testResolverId: String
}

type TestResolverDTOTestsCountAggregate {
  id: Int
  testResolverId: Int
}

type TestResolverDTOTestsMinMaxAggregate {
  id: ID
  testResolverId: String
}

type TestResolverDTOTestsAggregateResponse {
  groupBy: TestResolverDTOTestsAggregateGroupBy
  count: TestResolverDTOTestsCountAggregate
  distinctCount: TestResolverDTOTestsCountAggregate
  min: TestResolverDTOTestsMinMaxAggregate
  max: TestResolverDTOTestsMinMaxAggregate
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;

exports[`AggregateRelationsResolver aggregate should use the dtoName if provided new test 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
  relationsAggregate(
    \\"\\"\\"Filter to find records to aggregate on\\"\\"\\"
    filter: TestRelationDTOFilter

    \\"\\"\\"Limit the number of results group by aggregation can return\\"\\"\\"
    groupByLimit: Float
  ): [TestResolverDTORelationsAggregateResponse!]!
  testsAggregate(
    \\"\\"\\"Filter to find records to aggregate on\\"\\"\\"
    filter: TestRelationDTOFilter

    \\"\\"\\"Limit the number of results group by aggregation can return\\"\\"\\"
    groupByLimit: Float
  ): [TestResolverDTOTestsAggregateResponse!]!
}

input TestRelationDTOFilter {
  freeTextQuery: String
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
  contains: ID
  notContains: ID
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

type TestResolverDTORelationsAggregateGroupBy {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsCountAggregate {
  id: Int
  testResolverId: Int
}

type TestResolverDTORelationsMinMaxAggregate {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsAggregateResponse {
  groupBy: TestResolverDTORelationsAggregateGroupBy
  count: TestResolverDTORelationsCountAggregate
  distinctCount: TestResolverDTORelationsCountAggregate
  min: TestResolverDTORelationsMinMaxAggregate
  max: TestResolverDTORelationsMinMaxAggregate
}

type TestResolverDTOTestsAggregateGroupBy {
  id: ID
  testResolverId: String
}

type TestResolverDTOTestsCountAggregate {
  id: Int
  testResolverId: Int
}

type TestResolverDTOTestsMinMaxAggregate {
  id: ID
  testResolverId: String
}

type TestResolverDTOTestsAggregateResponse {
  groupBy: TestResolverDTOTestsAggregateGroupBy
  count: TestResolverDTOTestsCountAggregate
  distinctCount: TestResolverDTOTestsCountAggregate
  min: TestResolverDTOTestsMinMaxAggregate
  max: TestResolverDTOTestsMinMaxAggregate
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;

exports[`AggregateRelationsResolver aggregate should use the object type name 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
  relationsAggregate(
    \\"\\"\\"Filter to find records to aggregate on\\"\\"\\"
    filter: TestRelationDTOFilter

    \\"\\"\\"Limit the number of results group by aggregation can return\\"\\"\\"
    groupByLimit: Float
  ): [TestResolverDTORelationsAggregateResponse!]!
}

input TestRelationDTOFilter {
  freeTextQuery: String
  and: [TestRelationDTOFilter!]
  or: [TestRelationDTOFilter!]
  id: IDFilterComparison
  testResolverId: StringFieldComparison
}

input IDFilterComparison {
  is: Boolean
  isNot: Boolean
  eq: ID
  neq: ID
  gt: ID
  gte: ID
  lt: ID
  lte: ID
  like: ID
  notLike: ID
  iLike: ID
  notILike: ID
  in: [ID!]
  notIn: [ID!]
  contains: ID
  notContains: ID
}

input StringFieldComparison {
  is: Boolean
  isNot: Boolean
  eq: String
  neq: String
  gt: String
  gte: String
  lt: String
  lte: String
  like: String
  notLike: String
  iLike: String
  notILike: String
  in: [String!]
  notIn: [String!]
}

type TestResolverDTORelationsAggregateGroupBy {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsCountAggregate {
  id: Int
  testResolverId: Int
}

type TestResolverDTORelationsMinMaxAggregate {
  id: ID
  testResolverId: String
}

type TestResolverDTORelationsAggregateResponse {
  groupBy: TestResolverDTORelationsAggregateGroupBy
  count: TestResolverDTORelationsCountAggregate
  distinctCount: TestResolverDTORelationsCountAggregate
  min: TestResolverDTORelationsMinMaxAggregate
  max: TestResolverDTORelationsMinMaxAggregate
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;

exports[`AggregateRelationsResolver should not add read methods if one and many are empty 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;
